src/ophyd_async/epics/adaravis/_aravis.py

```diff
@@ -27,24 +24,28 @@ def __init__(
         hdf_suffix="HDF1:",
         name="",
         gpio_number: AravisController.GPIO_NUMBER = 1,
+        config_sigs: Sequence[SignalR] = (),
     ):
-        self.drv = AravisDriverIO(prefix + drv_suffix)
-        self.hdf = adcore.NDFileHDFIO(prefix + hdf_suffix)
-
         super().__init__(
-            AravisController(self.drv, gpio_number=gpio_number),
-            adcore.ADHDFWriter(
-                self.hdf,
-                path_provider,
-                lambda: self.name,
-                adcore.ADBaseDatasetDescriber(self.drv),
-            ),
-            config_sigs=(self.drv.acquire_time,),
+            prefix,
+            path_provider,
+            adcore.ADHDFWriter,
+            hdf_suffix,
+            AravisController,
+            AravisDriverIO,
+            drv_suffix=drv_suffix,
             name=name,
+            config_sigs=config_sigs,
+            gpio_number=gpio_number,
         )
+        self.hdf = self._fileio
```

Comment:
I'm not a big fan of inserting another class into the heirarchy with many more arguments, could we move out the creation of the driver and writer classes into a utility function instead? I had a go with this idea and came up with:
```python
class AravisDetectorTIFF(StandardDetector[AravisController]):
    def __init__(
        self,
        prefix: str,
        path_provider: PathProvider,
        drv_suffix="cam1:",
        tiff_suffix="TIFF1:",
        gpio_number: AravisController.GPIO_NUMBER = 1,
        config_sigs: Sequence[SignalR] = (),
        name="",
    ):
        self.drv, self.tiff, writer = adcore.areadetector_driver_and_tiff(
            drv_cls=AravisDriverIO,
            prefix=prefix,
            drv_suffix=drv_suffix,
            fileio_suffix=tiff_suffix,
            path_provider=path_provider,
        )
        super().__init__(
            controller=AravisController(self.drv, gpio_number),
            writer=writer,
            config_sigs=config_sigs,
            name=name,
        )
```

I did some of the changes required so pyright was happy, but didn't make the tests work:
https://github.com/bluesky/ophyd-async/commit/efd86c96cac8bcf9517f5a2555b5dc4acbfc50eb

Reply:
https://github.com/jwlodek/ophyd-async/pull/2

Reply:
Would it be possible to have the utility function(s) be factories that take in a drv_cls and writer_cls? It would be nice if we didn't need to make a function for each writer type. I am OK with that approach generally though

Reply:
That sounds reasonable, I wonder if we should stretch it even further and say that we make one `AravisDetector` that has an enum init arg `fileio_type` that switches between HDF and TIFF?

Reply:
That would be ideal I think. I'll work on getting something like that working.

Reply:
Possible we could even do something with a custom suffix, although this might be too magic
```python
class AravisDetector(StandardDetector[AravisController]):
    def __init__(
        self,
        prefix: str,
        path_provider: PathProvider,
        drv_suffix="cam1:",
        fileio_suffix: FileIOSuffix = NDFileHDFSuffix("HDF:")
        gpio_number: AravisController.GPIO_NUMBER = 1,
        config_sigs: Sequence[SignalR] = (),
        name="",
    ):
        self.drv, self.fileio, writer = adcore.areadetector_driver_and_writer(
            drv_cls=AravisDriverIO,
            prefix=prefix,
            drv_suffix=drv_suffix,
            fileio_suffix=fileio_suffix,
            path_provider=path_provider,
        )
        super().__init__(
            controller=AravisController(self.drv, gpio_number),
            writer=writer,
            config_sigs=config_sigs,
            name=name,
        )
```

Reply:
Hmmm, not sure I love that, but it would be nice to be able to specify the writer and have a default suffix. What if `fileio_suffix=None`, and 

```
if fileio_suffix is None:
    fileio_suffix = ad_writer_type.value
```

And have the enum values for the writer types just correspond to the default suffixes?

Reply:
Found another possible approach which avoids the need to make detector specific subclasses:
```python
ADBaseIOT = TypeVar("ADBaseIOT", bound=ADBaseIO)
NDFileIOT = TypeVar("NDFileIOT", bound=NDFileIO)
NDPluginBaseIOT = TypeVar("NDPluginBaseIOT", bound=NDPluginBaseIO)


class AreaDetector(
    Generic[ADBaseIOT, NDFileIOT, DetectorControllerT],
    StandardDetector[DetectorControllerT],
):
    def __init__(
        self,
        drv: ADBaseIOT,
        controller: DetectorControllerT,
        fileio: NDFileIOT,
        path_provider: PathProvider,
        plugins: dict[str, NDPluginBaseIO],
        config_sigs: Sequence[SignalR] = (),
        name: str = "",
    ):
        self.drv = drv
        self.fileio = fileio
        for name, plugin in plugins.items():
            setattr(self, name, plugin)

        def name_provider():
            return self.name

        if isinstance(fileio, NDFileHDFIO):
            writer = ADHDFWriter(
                fileio,
                path_provider,
                name_provider,
                ADBaseDatasetDescriber(drv),
                *plugins.values(),
            )
        else:
            writer = ADTIFFWriter(
                fileio, path_provider, name_provider, ADBaseDatasetDescriber(drv)
            )
        super().__init__(controller, writer, config_sigs, name)

    def get_plugin(
        self, name: str, plugin_type: type[NDPluginBaseIOT] = NDPluginBaseIO
    ) -> NDPluginBaseIOT:
        plugin = getattr(self, name, None)
        if not isinstance(plugin, plugin_type):
            raise TypeError(
                f"Expected {self.name}.{name} to be a {plugin_type}, got {plugin}"
            )
        return plugin
```
So now `det.drv` and `det.fileio` are always present, and we can do `det.get_plugin("stats", NDPluginStatsIO)` if we created it with `AreaDetector(..., plugins={"stats": NDPluginStatsIO(prefix+"STATS")}`

Reply:
So the detector specific classes turn into helper functions:
```python
def _aravis_drv_controller(
    prefix: str, drv_suffix: str, gpio_number: AravisController.GPIO_NUMBER
) -> tuple[AravisDriverIO, AravisController]:
    drv = AravisDriverIO(prefix + drv_suffix)
    controller = AravisController(drv, gpio_number)
    return drv, controller


def aravis_detector_hdf(
    prefix: str,
    path_provider: PathProvider,
    drv_suffix="cam1:",
    hdf_suffix="HDF1:",
    gpio_number: AravisController.GPIO_NUMBER = 1,
    plugins: dict[str, NDPluginBaseIO] | None = None,
    config_sigs: Sequence[SignalR] = (),
    name="",
) -> adcore.AreaDetector:
    drv, controller = _aravis_drv_controller(prefix, drv_suffix, gpio_number)
    fileio = adcore.NDFileHDFIO(prefix + hdf_suffix)
    return adcore.AreaDetector(
        drv, controller, fileio, path_provider, plugins or {}, config_sigs, name
    )


def aravis_detector_tiff(
    prefix: str,
    path_provider: PathProvider,
    drv_suffix="cam1:",
    tiff_suffix="TIFF:",
    gpio_number: AravisController.GPIO_NUMBER = 1,
    plugins: dict[str, NDPluginBaseIO] | None = None,
    config_sigs: Sequence[SignalR] = (),
    name="",
) -> adcore.AreaDetector:
    drv, controller = _aravis_drv_controller(prefix, drv_suffix, gpio_number)
    fileio = adcore.NDFileIO(prefix + tiff_suffix)
    return adcore.AreaDetector(
        drv, controller, fileio, path_provider, plugins or {}, config_sigs, name
    )
```

Reply:
Pushed again to https://github.com/jwlodek/ophyd-async/pull/2